### PR TITLE
Open-source multitenant concourse

### PR DESCRIPTION
**Pull request by tlwr**
See https://github.com/alphagov/re-autom8/pull/10

---

⬇️ pasted below some from ⬆️

What
---------

Creates modules for a multi-tenant Concourse managed via terraform, with monitoring/alerting via Prometheus/Grafana.

Why
--------

Managing one multitenant concourse is better than managing N single tenant ones

We have other Concourses which we manage which have divergent codebases.

xkcd 927

Concourse is built to be multitenant ([this medium post is extremely relevant](https://medium.com/concourse-ci/earning-our-wings-a0c307fa73e6)) and we can operate a single Concourse without compromising security (not for everything)

The design for this multitenant Concourse is incorporating and building upon the threat modelling / attack trees work that went into the other Concourse deployments.

How
--------

#### Architecture

Essentially there are 3 layers of isolation/permissions:
- Network level
- IAM level
- OAuth level

Concourse has a native concept called a `team` which we are effectively using as a role, e.g. `team1-reporting`, `team1-deploy`, `team1-monitor`. Access control is achieved using GitHub OAuth. Individual concourse workers can be scoped to an individual Concourse team, which provides isolation between teams.

Because Concourse workers can be scoped to a different team, we can provision different `pools` of workers using different autoscaling groups with different IAM instance profiles for different Concourse teams (i.e. different roles, e.g. `team1-x` instances have a different IAM role to `team2-x` instances).

We also make use of IP whitelisting, it would be unusable if `team1-x` and `team2-x` shared the same egress IPs, because we would have no network isolation. For convenience sake I have introduced the concept of `network pools` which are NAT gateways with static IPs which are the egress gateways for ONE OR MORE `worker pools`. I.e. `verify-deploy` and `verify-monitor` could share the same egress IP (if we wanted them to).

#### Security and secrets

Concourse has 2 high level components, `web` nodes which are shared amongst all tenants, these are used for scheduling and the web UI, and `worker` nodes which we scope to individual Concourse teams (roles).

Web instances can access all secrets (because they schedule the jobs and need to unlock the secrets to do so), however tenant users cannot access secrets other than running a task in a pipeline (which only has access to the team's secrets).

Worker instances (by choice not inherent design limitations) can also access their own secrets (we should probably namespace these to writeable secrets and unwriteable secrets), this is so tenants can manage their own secrets (I envision a `provision` pipeline managed by the tenants which looks at a git repo for which secrets are encrypted to a public key accessible by the worker instances, for provisioning secrets, but this is out of scope of this PR).

Because IAM roles isolate team/roles from each other we have nice isolation that way.

#### Network topology

Other than previously described, please read this [Pivotal architecture doc about Concourse](https://docs.pivotal.io/p-concourse/4-x/architecture.html)

Workers cannot access each other. Web nodes can access each other (they need to). Everything has egress to the internet as there are no egress proxies.

#### Monitoring

In the environment there are essentially 3 user facing components:

- ✈️ Concourse web UI
- 📈 Grafana
- 🔥 Prometheus

Grafana and Prometheus are part of the `monitoring` module so are optional. However node-exporter runs on every instance, and concourse-web is scrapeable regardless.

Prometheus scrapes all the instances in the environment using ec2 service discovery according to the `Deployment` EC2 instance tag. You can use Prometheus for running ad-hoc queries but it isn't actually very nice for this.

Grafana (version 5.4) is running in the environment and frontend by OAuth (github). Although right now the permissions are managed manually (i.e. ❄️ through the web UI) - this can be done with Grafana terraform provider.

Grafana has the new `explore` feature enabled which will be enabled by default in Grafana 6. The explore feature is IMO better than Prometheus for exploring Prometheus metrics.

image
*an image showing Grafana's native explore feature as a replacement for the Prometheus UI*

In a private deployment I have ❄️ manually added a dashboard with some helpful alerts (which don't go anywhere yet) about the health of the various parts of the deployment.

screen shot 2019-01-05 at 15 52 37
*an image showing an initial dashboard/alerting pane for the cd Concourse deployment*

---

Further work
-------------

- Perhaps disable external access to Prometheus as it does not seem to add much value.

- S3 buckets per tenant. Due to the isolation between IAM roles we could create 2 "magic" s3 buckets per tenant which could be managed by the instances. E.g. a public bucket and a private bucket, this would alleviate the need for people to do a lot of assume role calls for Concourse resources. These would be scoped to individual tenants obviously.

- ECR repositories per tenant. As part of a provisioning job (above) tenants could be able to provision their own ECR repositories for their docker images. The worker instances for each role would only be able to manage the buckets in their namespace so we would have nice isolation between tenants.

- Tagging. We can improve the tagging of resources created on behalf of each tenant, not only would this help us navigate around the AWS account, but it would also be useful in the future if we decide to do chargebacks.

- Continuous deployment. We can get the `main` concourse team (i.e. global admin) to continuously deploy the cluster (or at least the operational bits - i.e. rolling worker nodes, etc).

- Grafana HA. Right now Grafana is not HA (due to how Grafana does sessions). We should make Grafana HA so config changes are ZDD.

- Prometheus HA.

- Create an example deployment so other open source people can see how to deploy it, or internal users for testing.

### CODE REVIEW

File: reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        target_label: instance
+
+  - job_name: concourse_web
+    ec2_sd_configs:
+      - region: eu-west-2
+        refresh_interval: 30s
+        port: 9100
+      - region: eu-west-2
+        refresh_interval: 30s
+        port: 9101
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh]: all [instances](https://prometheus.io/docs/concepts/jobs_instances/) under the same job name should expose the same metric. different ports generally means different metrics and so should be exposed under different job names.  (according to the [default port allocations](https://github.com/prometheus/prometheus/wiki/Default-port-allocations), 9100 is node_exporter and 9101 is haproxy_exporter?)

File: reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        refresh_interval: 30s
+        port: 9100
+    relabel_configs:
+      - source_labels: [__meta_ec2_tag_Name]
+        regex: '^${deployment}-.*-concourse-worker$'
+        action: keep
+      - source_labels: [__meta_ec2_instance_id]
+        target_label: instance
+
+  - job_name: concourse_all
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh]: I think this should probably be called `concourse_node`?

File: reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    ec2_sd_configs:
+      - region: eu-west-2
+        refresh_interval: 30s
+        port: 9100
+    relabel_configs:
+      - source_labels: [__meta_ec2_instance_id]
+        target_label: instance
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/files/prometheus-init.sh]: is it worth adding target labels to identify what kind of instance we're dealing with here? you can populate prometheus labels from AWS tags.

File: README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# alphagov/tech-ops
```
Comment (philandstuff) [inline README.md]: cool readme m8

File: reliability-engineering/terraform/modules/concourse-monitoring/lb.tf [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    Deployment = "${var.deployment}"
+  }
+}
+
+resource "aws_lb_listener" "concourse_monitoring_https" {
+  load_balancer_arn = "${aws_lb.concourse_monitoring.arn}"
+  port              = "443"
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-2015-05"
+  certificate_arn   = "${aws_acm_certificate.concourse_monitoring.arn}"
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/lb.tf]:
```suggestion
  certificate_arn   = "${aws_acm_certificate_validation.concourse_web.certificate_arn}"
```
I don't see much point in having the `aws_acm_certificate_validation` resource unless you get the arn from it like this.  This is what actually causes terraform to wait for validation before attaching it to the load balancer.

Note that this is much more important when making changes to infrastructure (eg changing the certificate) than in the initial buildout.

File: reliability-engineering/terraform/modules/concourse-monitoring/data.tf [hunk 6]
```diff
@@ -0,0 +1,7 @@
+data "aws_vpc" "concourse" {
+  id = "${var.vpc_id}"
+}
+
+locals {
+  vpc_dns_resolver = "${cidrhost(data.aws_vpc.concourse.cidr_block, 2)}"
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/data.tf]: ooh i didn't know this trick, nice 👍

File: reliability-engineering/terraform/modules/concourse-monitoring/grafana.tf [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  username                  = "grafana"
+  password                  = "${random_string.concourse_grafana_db_password.result}"
+  vpc_security_group_ids    = ["${aws_security_group.concourse_grafana_db.id}"]
+  db_subnet_group_name      = "${aws_db_subnet_group.concourse_grafana_db.name}"
+  final_snapshot_identifier = "${var.deployment}-concourse-grafana-final"
+  backup_retention_period   = 7
+  storage_encrypted         = true
+}
+
+data "template_file" "concourse_grafana_task_def" {
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/grafana.tf]: really minor thing, but this isn't a task definition, it's a container definition

File: reliability-engineering/terraform/modules/concourse-monitoring/lb.tf [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+  default_action {
+    type = "fixed-response"
+
+    fixed_response {
+      content_type = "text/plain"
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-monitoring/lb.tf]:
```suggestion
      content_type = "text/plain; charset=utf-8"
```
don't want mojibake to ruin the emoji 😁

File: reliability-engineering/terraform/modules/concourse-web/lb.tf [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  protocol = "HTTP"
+  vpc_id   = "${var.vpc_id}"
+}
+
+resource "aws_lb_listener" "concourse_web_https" {
+  load_balancer_arn = "${aws_lb.concourse_web.arn}"
+  port              = "443"
+  protocol          = "HTTPS"
+  ssl_policy        = "ELBSecurityPolicy-2015-05"
+  certificate_arn   = "${aws_acm_certificate.concourse_public_deployment.arn}"
```
Comment (philandstuff) [inline reliability-engineering/terraform/modules/concourse-web/lb.tf]: ditto `aws_acm_certificate_validation`

File: reliability-engineering/terraform/modules/concourse-monitoring/lb.tf (hunk 5 above)
Comment (tlwr) [reply reliability-engineering/terraform/modules/concourse-monitoring/lb.tf]: I use the resource to generate the DNS records. This is a good change though.

Comment (philandstuff) [reply reliability-engineering/terraform/modules/concourse-monitoring/lb.tf]: > I use the resource to generate the DNS records. 

but that's... not what `aws_acm_certificate_validation` does? And you still create the DNS records yourself anyway? 😕

**Comment by tlwr** [comment]
❌ emoji content type, i've backed this out because it is not supported by AWS ALBs

✅ ACM cert validation certificate_arn

**Pull request merged**